docs/adr/0008-secrets-management.md

```diff
@@ -0,0 +1,53 @@
+- Encrypted manifests should be compatible with Kustomize.
+- Ideally only the specific portions within manifests that are confidential will be encrypted (e.g. specific fields).
+- ArgoCd should be able to decrypt these secrets and thus deploy them.
+- Other teams can easily encrypt their manifests whilst maintaining ease of consumption by ArgoCD.
+- Decryption capabilities are limited to code-owners, operate-first admins, and ArgoCD.
+
+## Considered Options
+
+There exist many different solutions to tackle GitOps with secret management. ArgoCD has a list of recommended tools one can leverage, you may find it [here](https://argoproj.github.io/argo-cd/operator-manual/secret-management/). We considered the following tools:
+
+1) [Sealed Secrets](https://github.com/bitnami-labs/sealed-secrets) - An operator that can decrypt a secret from a custom resource, this resource is stored in Git.
+2) [Hashicorp Vault](https://www.vaultproject.io/) - A server with a backing storage where secrets are stored.
+3) [KSOPS](https://github.com/viaduct-ai/kustomize-sops#argo-cd-integration) - A Kustomize plugin that uses [sops](https://github.com/mozilla/sops) to encrypt, the plugin itself adds a decryption generator step to the Kustomize build.
+4) Store secrets in a private/internal repo
+
+## Decision Outcome
+Chosen Option `(3)`. Reasons are summarized as follows:
+
+- `Sealed Secrets` only allow you to encrypt k8s `secrets`
+- `Hashicorp Vault` a bit too heavy duty for our use case, steep learning curve. Secrets are also stored in a backend storage and not in Git
+- We would like to avoid storing manifests in private/internal repos. This is to stay in line with keeping our system's desired state as public as possible.
+
+By using KSOPS we can encrypt specific portions of a manifest that are confidential whilst keeping the rest of the non-confidential information visible. Encrypted manifests can be stored in whichever location as needed. By using an ArgoCD image that has KSOPS installed, we can have ArgoCD also decrypt these manifests.
+
+SOPS allows you to use various methods to encrypt your manifests, we opt to use GPG as it's free to generate and use (other methods include using AWS KMS, GCP KMS, and Azure Key Vault).
+
+We will create a GPG key that we will use to encrypt all our manifests in Git. The secret key will be made available to ArgoCD and other operate-first admins. We offer the public key to other teams should they wish to encrypt their contents and store them in Git. Doing so will allow ArgoCD to decrypt and deploy their manifests. This means that other teams will be able to encrypt, but not decrypt, their manifests in Git. To get around this, teams will be able to create their own GPG keys and make a request that only our ArgoCD be given this new private key. This way, teams can encrypt manifests in their repo, while retaining the ability to decrypt them as well.
```

Comment:
> This means that other teams will be able to encrypt, but not decrypt, their manifests in Git. To get around this, teams will be able to create their own GPG keys and make a request that only our ArgoCD be given this new private key. This way, teams can encrypt manifests in their repo, while retaining the ability to decrypt them as well.

Not sure about this. I'd rather see them using their own GPG key and then specifying it in the `.sops` file, so each specific team owned resource is encrypted by 2 keys - OperateFirst admin + team owned. Both keys can encrypt and decrypt, so each team can also modify resources while the admin key maintains access to the resource. This is basically the setup we used in https://github.com/aicoe-aiops/project-template/blob/master/manifests/.sops.yaml

Reply:
I'm fine with that. Admins would technically have access to the private key required to decrypt the team's manifests in either case (in the former, since we need to put it in argocd), but with your suggestion, it keeps it a bit formalized, I'll update.

Reply:
I've updated it, this is a significantly better way to do it, good call.